substrate/primitives/core/src/crypto_bytes.rs

```diff
@@ -0,0 +1,240 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Generic byte array which can be specialized with a marker type.
+
+use crate::{
+	crypto::{FromEntropy, UncheckedFrom},
+	hash::{H256, H512},
+};
+
+use codec::{Decode, Encode, MaxEncodedLen};
+use core::marker::PhantomData;
+use scale_info::TypeInfo;
+use sp_runtime_interface::pass_by::{self, PassBy, PassByInner};
+
+/// Generic byte array.
+///
+/// The type is generic over a constant length `N` and a "tag" `T` which
+/// can be used to specialize the byte array without using newtypes.
+#[derive(Encode, Decode, MaxEncodedLen)]
+#[repr(transparent)]
+pub struct CryptoBytes<const N: usize, T = ()>(pub [u8; N], PhantomData<fn() -> T>);
```

Comment:
Some code comment explaining the PhantomData and why it looks like this. I forgot the actual name of this and also how it actually unfolds. I only remember that it changes if it is sync or send etc.

There exists some doc page in rust, maybe worthwhile to link this one.

Reply:
https://github.com/paritytech/polkadot-sdk/pull/3684/commits/cd089eabd554fc85853dc0a39b659f6a60b61f11